webapp/src/components/ImageEditor/Timeline.tsx

```diff
@@ -0,0 +1,327 @@
+            const { currentFrame, dispatchChangeCurrentFrame } = this.props;
+            if (this.dragEnd) this.dragEnd = false;
+            else if (index != currentFrame) dispatchChangeCurrentFrame(index);
+        };
+
+        return this.handlers[index];
+    }
+
+    protected duplicateFrame = () => {
+        const { currentFrame, dispatchDuplicateFrame } = this.props;
+        dispatchDuplicateFrame(currentFrame);
+    }
+
+    protected deleteFrame = () => {
+        const { currentFrame, dispatchDeleteFrame } = this.props;
+        dispatchDeleteFrame(currentFrame);
+    }
+
+    protected newFrame = () => {
+        const { dispatchNewFrame } = this.props;
+        if (this.dragEnd) this.dragEnd = false;
+        else dispatchNewFrame();
+    }
+
+    protected redraw() {
+        if (!this.canvas) return;
+
+        const { frames, currentFrame } = this.props;
+
+        const imageState = frames[currentFrame];
+
+        this.canvas.height = imageState.bitmap.height;
+        this.canvas.width = imageState.bitmap.width;
+
+        const bitmap = Bitmap.fromData(imageState.bitmap);
+        this.drawBitmap(bitmap);
+
+        if (imageState.floatingLayer) {
+            const floating = Bitmap.fromData(imageState.floatingLayer);
+            this.drawBitmap(floating, imageState.layerOffsetX, imageState.layerOffsetY, true);
+        }
+    }
+
+    protected drawBitmap(bitmap: Bitmap, x0 = 0, y0 = 0, transparent = false) {
+        const { colors } = this.props;
+
+        const context = this.canvas.getContext("2d");
+        context.imageSmoothingEnabled = false;
+        for (let x = 0; x < bitmap.width; x++) {
+            for (let y = 0; y < bitmap.height; y++) {
+                const index = bitmap.get(x, y);
+
+                if (index) {
+                    context.fillStyle = colors[index];
+                    context.fillRect(x + x0, y + y0, 1, 1);
+                }
+                else {
+                    if (!transparent) context.clearRect(x + x0, y + y0, 1, 1);
+                }
+            }
+        }
+    }
+
+    protected updateDragDrop(coord: ClientCoordinates) {
+        const parent = this.frameScroller.getBoundingClientRect();
+        const scrollY = coord.clientY - parent.top;
+
+        const unit = this.frameScroller.firstElementChild.getBoundingClientRect().height;
+        const index = Math.floor(scrollY / unit);
+
+        if (!this.state.isMovingFrame) {
+            this.setState({
+                isMovingFrame: true,
+                dropPreviewIndex: this.props.currentFrame,
+            });
+        }
+        else if (this.refs["floating-frame"]) {
+            const floating = this.refs["floating-frame"] as HTMLDivElement;
+            floating.style.top = scrollY + this.scrollOffset + "px";
+
+            this.setState({ dropPreviewIndex: index });
+        }
+    }
+}
+
+export function addClass(el: SVGElement | HTMLElement, classes: string) {
```

Comment:
do these need to be copied in here? if so can we add a comment pointing to the original definitions to make sure they stay in sync

Reply:
Nope, this was an artifact that should have been removed. Will do